rfcs/20190722-tflite-training.md

```diff
@@ -0,0 +1,427 @@
+
+### Using the Model in TensorFlow Lite runtime
+
+There's an ongoing design effort to add enable users to call different subgraphs
+in a TensorFlow Lite model. Tentatively the usage may look like:
+
+```
+Interpreter interpreter = Interpreter(...);
+{
+  auto subgraph = interpreter.SelectSubgraph("train_dense_layer");
+  subgraph.SetInputTensor(0, training_feature);
+  subgraph.SetInputTensor(1, training_label);
+  subgraph.Invoke();
+}
+{
+  auto subgraph = interpreter.SelectSubgraph("inference");
+  subgraph.SetInputTensor(0, inference_feature);
+  subgraph.Invoke();
+  auto result = subgraph.GetOutputTensor(0);
+}
+```
+
+Regardless of how the new API design look like, users should be able to choose
+to run `train_dense_layer` or `inference` subgraph.
+
+### Testing TensorFlow Lite Model correctness
+
+On-device training is complex and it can be hard to troubleshoot. TensorFlow
+Lite should provide guidelines and tools to make this easy. The core idea is: If
+each TensorFlow function is converted to a TensorFlow Lite subgraph without
+changing semantics, it will be easy to test TensorFlow and TensorFlow Lite
+functionality side by side.
+
+In addition to the common ML testing best practices, there are a few specific
+suggestions for TensorFlow Lite on-device learning:
+
+*   **Test all functions in TensorFlow** before converting to TensorFlow Lite.
+    In this use case, in addition to testing `inference` and `train`, also test
+    if `train_dense_layers` since it's used on device. If it doesn't work well
+    in TensorFlow, it can't possibly work well in TensorFlow Lite. Check if
+    there are any coding error, or you may need to tune the model structure or
+    hyperparameter.
+*   **Use model diff testing tool to test all functions**. TensorFlow Lite now
+    has a tool which feeds the same random data into TensorFlow and TensorFlow
+    Lite, and compare if the results are close enough in a threshold. The model
+    should be extended to handle multiple subgraphs (including training and
+    inference subgraphs) and variables. This is an end to end test which can
+    capture kernel bugs or TensorFlow/TensorFlow Lite discrepancy.
+
+## Implementation details for basic training
+
+There is nothing magical in training. Technically, training is doing a series of
+**mathematical computation** (requiring these op kernels), and update the weight
+**variables** which is **accessible by other (inference) subgraphs**. The
+converter also needs to be changed, to support and optimize training use cases.
+
+### Sharing variables across subgraphs
+
+For the variable requirement of TensorFlow Lite training, the high level goals
+are:
+
+*   Convert TensorFlow Resource Variables to TensorFlow Lite, and preserve the
+    same semantic
+*   Able to share variables between multiple TensorFlow Lite subgraphs
+
+We propose to define variables in TFLite that have similar semantics to resource
+variables in TensorFlow:
```

Comment:
Why do we need a special variable type? Can you explain further what the difference is?

Reply:
Currently TensorFlow Lite doesn't really(*) support variables. Essentially there's no way to update the trainable weights. That's why we need a new design. 

(*) In TensorFlow Lite codebase, there's a concept called "variable", but it's only works with fused RNN ops now and can't be shared across subgraphs. It doesn't fulfill the requirement for training. In the long term we want to deprecate the old variable and migrate to the new resource-variable like semantic.